File: backend/app/schemas/ai.py
"""API-level schemas for AI analysis and similar-incident endpoints."""
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel

from app.ai.schemas import AIAnalysisResponse
from app.models.ai_analysis import ConfidenceInfo, EvidenceRef


class AnalyzeSuccessResponse(BaseModel):
    status: Literal["ok"] = "ok"
    analysis_id: str
    incident_id: str
    model: str
    prompt_version: str
    analysis: AIAnalysisResponse
    evidence: list[EvidenceRef]
    confidence: ConfidenceInfo
    retrieval_count: int
    latency_ms: int
    created_at: datetime


class AnalyzeUnavailableResponse(BaseModel):
    status: Literal["ai_unavailable"] = "ai_unavailable"
    message: str
    retryable: bool = True


class SimilarIncidentItem(BaseModel):
    incident_id: str
    title: str
    similarity: float
    relationship: Literal["duplicate", "related"]
    status: Optional[str] = None
    resolution_summary: Optional[str] = None


class SimilarIncidentsResponse(BaseModel):
    incident_id: str
    duplicate_threshold: float
    related_threshold: float
    items: list[SimilarIncidentItem]

File: frontend/src/types/ai.ts
export type ConfidenceLevel = "High" | "Medium" | "Low";

export interface ProbableCause {
  cause: string;
  evidence_ids: string[];
  likelihood: ConfidenceLevel;
}

export interface RecommendedAction {
  action: string;
  evidence_ids: string[];
  priority_order: number;
}

export interface SimilarIncidentMention {
  incident_id: string;
  evidence_id: string;
  relationship: "duplicate" | "related";
  rationale?: string | null;
}

export interface KnowledgeArticleMention {
  article_id: string;
  evidence_id: string;
  relevance?: string | null;
}

export interface AIAnalysisResponse {
  summary: string;
  category: string;
  priority: "P1" | "P2" | "P3" | "P4";
  probable_causes: ProbableCause[];
  recommended_actions: RecommendedAction[];
  similar_incidents: SimilarIncidentMention[];
  knowledge_articles: KnowledgeArticleMention[];
  escalation_required: boolean;
  confidence: ConfidenceLevel;
  uncertainties: string[];
  final_recommendation: string;
}

export interface EvidenceRef {
  evidence_id: string;
  document_type: "knowledge" | "historical-tickets";
  document_id: string;
  title?: string | null;
  score?: number | null;
  source?: string | null;
}

export interface ConfidenceInfo {
  model_reported?: string | null;
  evidence_score: number;
  bucket: ConfidenceLevel;
}

export interface AnalyzeSuccessResponse {
  status: "ok";
  analysis_id: string;
  incident_id: string;
  model: string;
  prompt_version: string;
  analysis: AIAnalysisResponse;
  evidence: EvidenceRef[];
  confidence: ConfidenceInfo;
  retrieval_count: number;
  latency_ms: number;
  created_at: string;
}

export interface AnalyzeUnavailableResponse {
  status: "ai_unavailable";
  message: string;
  retryable: boolean;
}

export type AnalyzeResponse = AnalyzeSuccessResponse | AnalyzeUnavailableResponse;

export interface SimilarIncidentItem {
  incident_id: string;
  title: string;
  similarity: number;
  relationship: "duplicate" | "related";
  status?: string | null;
  resolution_summary?: string | null;
}

export interface SimilarIncidentsResponse {
  incident_id: string;
  duplicate_threshold: number;
  related_threshold: number;
  items: SimilarIncidentItem[];
}
